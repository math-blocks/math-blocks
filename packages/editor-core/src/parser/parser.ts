import {UnreachableCaseError} from "@math-blocks/core";
import * as Parser from "@math-blocks/parser-factory";
import * as Semantic from "@math-blocks/semantic";
import type {Mutable} from "utility-types";

import * as Lexer from "./lexer";
import {locFromRange} from "../token/util";
import {parseVerticalWork} from "./vertical-work";

import type {CharRow} from "../char/types";
import type {TokenNode, SourceLocation} from "../token/types";

// TODO: fill out this list
type Operator =
    | "add"
    | "sub"
    | "plusminus"
    | "mul.exp"
    | "div"
    | "mul.imp"
    | "neg"
    | "eq"
    | "supsub"
    | "nul";

type NAryOperator = "add" | "sub" | "plusminus" | "mul.exp" | "mul.imp" | "eq";

type EditorParser = Parser.IParser<TokenNode, Parser.types.Node, Operator>;

const isIdentifier = (node: TokenNode): boolean =>
    node.type === "token" && node.name === "identifier";

const getPrefixParselet = (
    node: TokenNode,
): Parser.PrefixParselet<TokenNode, Parser.types.Node, Operator> => {
    switch (node.type) {
        case "token": {
            switch (node.name) {
                case "identifier":
                    return {
                        parse: () =>
                            Parser.builders.identifier(node.value, node.loc),
                    };
                case "number":
                    return {
                        parse: () =>
                            Parser.builders.number(node.value, node.loc),
                    };
                case "minus":
                    return {
                        parse: (parser) => {
                            const neg = parser.parseWithOperator("neg");
                            const loc = locFromRange(node.loc, neg.loc);
                            return Parser.builders.neg(neg, false, loc);
                        },
                    };
                case "plusminus":
                    return {
                        parse: (parser) => {
                            const neg = parser.parseWithOperator("plusminus");
                            const loc = locFromRange(node.loc, neg.loc);
                            return Parser.builders.plusminus(neg, "unary", loc);
                        },
                    };
                case "ellipsis":
                    return {
                        parse: () => Parser.builders.ellipsis(node.loc),
                    };
                default:
                    throw new Error(`Unexpected '${node.name}' atom`);
            }
        }
        case "frac":
            return {
                parse: () => {
                    const [numerator, denominator] = node.children;
                    return Parser.builders.div(
                        editorParser.parse(numerator.children),
                        editorParser.parse(denominator.children),
                        node.loc,
                    );
                },
            };
        case "root":
            return {
                parse: () => {
                    const [index, radicand] = node.children;
                    return index === null
                        ? Parser.builders.sqrt(
                              editorParser.parse(radicand.children),
                              node.loc,
                          )
                        : Parser.builders.root(
                              editorParser.parse(radicand.children),
                              editorParser.parse(index.children),
                              node.loc,
                          );
                },
            };
        case "delimited":
            return {
                parse: () => {
                    const [inner] = node.children;
                    const result = editorParser.parse(inner.children);
                    // TODO: what should `loc` be here?
                    return Parser.builders.parens(result);
                },
            };
        case "table":
            if (node.subtype === "algebra") {
                return {
                    parse: () => {
                        return parseVerticalWork(node, editorParser);
                    },
                };
            }
            throw new Error("We don't handle 'table' tokens yet");
        // TODO: Handle subsup at the start of a row, useful in Chemistry
        case "subsup":
            throw new Error("Unexpected 'subsup' token");
        // TODO: Handle limits at the start of a row
        case "limits":
            throw new Error("Unexpected 'limits' token");
        case "row":
            throw new Error("Unexpected 'row' token");
        default:
            throw new UnreachableCaseError(node);
    }
};

// most (all?) of the binary only operations will be handled by the editor
// const parseBinaryInfix = (op: Operator) => (
//     parser: EditorParser,
//     left: Node,
// ): Node => {
//     parser.consume();
//     return {
//         type: op,
//         args: [left, parser.parseWithPrecedence(parser.getOpPrecedence(op))],
//     };
// };

const parseNaryInfix =
    (op: NAryOperator) =>
    (parser: EditorParser, left: Parser.types.Node): Parser.types.Node => {
        const [right, ...rest] = parseNaryArgs(parser, op);
        const loc = locFromRange(
            left.loc,
            rest.length > 0 ? rest[rest.length - 1].loc : right.loc,
        );

        switch (op) {
            case "add":
            case "sub":
            case "plusminus":
                return Parser.builders.add([left, right, ...rest], loc);
            case "mul.imp":
                return Parser.builders.mul([left, right, ...rest], true, loc);
            case "mul.exp":
                return Parser.builders.mul([left, right, ...rest], false, loc);
            case "eq":
                return Parser.builders.eq([left, right, ...rest], loc);
        }
    };

/**
 * Returns an array or one or more nodes that are arguments for the given
 * operator.  All n-ary operators require at least two arguments, but the
 * first argument is already parsed by parseNaryInfix so it makes sense
 * that the return value is one or more.
 */
const parseNaryArgs = (
    parser: EditorParser,
    op: NAryOperator,
): OneOrMore<Parser.types.Node> => {
    // TODO: handle implicit multiplication
    const node = parser.peek();
    if (node.type === "token") {
        if (node.name === "identifier" || node.name === "number") {
            // implicit multiplication
        } else {
            // an explicit operation, e.g. plus, times, etc.
            parser.consume();
        }
        let expr = parser.parseWithOperator(op);
        if (op === "sub") {
            const loc = locFromRange(node.loc, expr.loc);
            expr = Parser.builders.neg(expr, true, loc);
        }
        if (op === "plusminus") {
            const loc = locFromRange(node.loc, expr.loc);
            expr = Parser.builders.plusminus(expr, "binary", loc);
        }
        const nextToken = parser.peek();
        if (nextToken.type !== "token") {
            throw new Error("atom expected");
        }
        if (
            (op === "add" || op === "sub" || op === "plusminus") &&
            (nextToken.name === "plus" ||
                nextToken.name === "minus" ||
                nextToken.name === "plusminus")
        ) {
            if (nextToken.name === "plus") {
                op = "add";
            } else if (nextToken.name === "minus") {
                op = "sub";
            } else if (nextToken.name === "plusminus") {
                op = "plusminus";
            } else {
                throw new Error("unexpected value for nextAtom.kind");
            }
            return [expr, ...parseNaryArgs(parser, op)];
        } else if (op === "mul.exp" && nextToken.name === "times") {
            return [expr, ...parseNaryArgs(parser, op)];
        } else if (op === "mul.imp" && nextToken.name === "identifier") {
            return [expr, ...parseNaryArgs(parser, op)];
        } else if (op === "eq" && nextToken.name === "eq") {
            return [expr, ...parseNaryArgs(parser, op)];
        } else {
            return [expr];
            // TODO: deal with frac, subsup, etc.
        }
    } else if (node.type === "root") {
        parser.consume();
        const [index, radicand] = node.children;
        const expr =
            index === null
                ? Parser.builders.sqrt(
                      editorParser.parse(radicand.children),
                      node.loc,
                  )
                : Parser.builders.root(
                      editorParser.parse(radicand.children),
                      editorParser.parse(index.children),
                      node.loc,
                  );
        const nextToken = parser.peek();
        if (nextToken.type === "root" || isIdentifier(nextToken)) {
            return [expr, ...parseNaryArgs(parser, "mul.imp")];
        } else {
            return [expr];
        }
    } else if (node.type === "frac") {
        parser.consume();
        const [num, den] = node.children;
        const expr = Parser.builders.div(
            editorParser.parse(num.children),
            editorParser.parse(den.children),
            node.loc,
        );
        return [expr];
    } else if (node.type === "delimited") {
        parser.consume();
        const [inner] = node.children;
        // TODO: make 'eol' its own token type instead of a co-opting 'atom'
        const nextToken = parser.peek();
        const expr = Parser.builders.parens(editorParser.parse(inner.children));

        return nextToken.type === "delimited"
            ? [expr, ...parseNaryArgs(parser, op)]
            : [expr];
    } else {
        throw new Error(`we don't handle ${node.type} tokens yet`);
        // TODO: deal with frac, subsup, etc.
    }
};

const getInfixParselet = (
    node: TokenNode,
): Parser.InfixParselet<TokenNode, Parser.types.Node, Operator> | null => {
    switch (node.type) {
        case "token": {
            switch (node.name) {
                case "plus":
                    return {op: "add", parse: parseNaryInfix("add")};
                case "minus":
                    return {op: "add", parse: parseNaryInfix("sub")};
                case "plusminus":
                    return {
                        op: "plusminus",
                        parse: parseNaryInfix("plusminus"),
                    };
                case "times":
                    return {op: "mul.exp", parse: parseNaryInfix("mul.exp")};
                case "eq":
                    return {op: "eq", parse: parseNaryInfix("eq")};
                case "identifier":
                    return {op: "mul.imp", parse: parseNaryInfix("mul.imp")};
                case "number":
                    return {op: "mul.imp", parse: parseNaryInfix("mul.imp")};
                default:
                    return null;
            }
        }
        case "subsup": {
            // TODO: we need to look the previous node so we know if we should
            // be generating a sum or product node or an exponent node.  It also
            // means we have to replace the current last.  It's essentially a
            // postfix operator like ! (factorial).
            // TODO: determine the "op" based on what left is, but we can't currently do that
            return {
                op: "supsub",
                parse: (parser: EditorParser, left: Parser.types.Node) => {
                    parser.consume(); // consume the subsup
                    const [sub, sup] = node.children;

                    if (left.type === "Identifier") {
                        if (sub) {
                            left = {
                                ...left,
                                subscript: editorParser.parse(sub.children),
                            };
                        }
                    } else {
                        if (sub) {
                            throw new Error(
                                "subscripts are only allowed on identifiers",
                            );
                        }
                    }

                    if (sup) {
                        const loc = locFromRange(
                            left.loc,
                            left.loc,
                        ) as Mutable<SourceLocation>;
                        if (loc) {
                            // Add 1 to account for the subsup itself since left
                            // is the node the supsub is being applied to
                            loc.end += 1;
                        }

                        return Parser.builders.pow(
                            left,
                            editorParser.parse(sup.children),
                            loc,
                        );
                    }

                    return left;
                },
            };
        }
        case "root": {
            return {op: "mul.imp", parse: parseNaryInfix("mul.imp")};
        }
        case "frac": {
            return {
                op: "mul.imp",
                parse: (parser, left): Parser.types.Node => {
                    const parselet = parseNaryInfix("mul.imp");
                    if (left.type === "div") {
                        throw new Error(
                            "An operator is required between fractions",
                        );
                    }
                    return parselet(parser, left);
                },
            };
        }
        case "delimited": {
            return {
                // TODO: figure out how to return a different value for 'op' if
                // the delimited node stands for something else like function
                // arguments.
                op: "mul.imp",
                parse: (parser, left): Parser.types.Node => {
                    const parselet = parseNaryInfix("mul.imp");
                    // TODO: check the left.type can be implicitly multiplied
                    // with parens.
                    return parselet(parser, left);
                },
            };
        }
        case "table":
            throw new Error("We don't handle 'table' tokens yet");
        case "limits":
            throw new Error(`Unexpected 'limits' token`);
        case "row":
            throw new Error(`Unexpected 'row' token`);
        default:
            throw new UnreachableCaseError(node);
    }
};

const getOpPrecedence = (op: Operator): number => {
    switch (op) {
        case "nul":
            return 0;
        case "eq":
            return 2;
        case "add":
        case "sub":
        case "plusminus":
            return 3;
        case "mul.exp":
            return 5;
        case "div": // this is to encourage wrapping fractions in parens before a negative
            return 6;
        case "neg":
            return 7;
        case "mul.imp":
            return 8;
        case "supsub":
            return 10;
    }
};

const EOL: TokenNode = Lexer.atom(
    {type: "token", name: "eol"},
    Lexer.location([], -1, -1),
);

export const editorParser = Parser.parserFactory<
    TokenNode,
    Parser.types.Node,
    Operator
>(getPrefixParselet, getInfixParselet, getOpPrecedence, EOL);

// WARNING: This function mutates `node`.
const removeExcessParens = (node: Semantic.types.Node): Semantic.types.Node => {
    const path: Semantic.types.Node[] = [];

    return Semantic.util.traverse(node, {
        enter: (node) => {
            path.push(node);
        },
        exit: (node) => {
            path.pop();
            const parent = path[path.length - 1];
            if (!parent) {
                return;
            }

            // TODO: use the precedence of the operators to determine whether
            // the parens are necessary or not.
            if (node.type === "Parens") {
                const {arg} = node;
                if (parent.type === "Parens") {
                    return;
                }
                if (parent.type === "mul" && parent.implicit) {
                    return arg;
                }
                if (arg.type === "Identifier" || arg.type === "number") {
                    return;
                }
                if (arg.type === "mul" && parent.type === "add") {
                    return;
                }
                if (arg.type === "neg" && parent.type !== "pow") {
                    return;
                }
                return arg;
            }
        },
    });
};

export const parse = (input: CharRow): Semantic.types.Node => {
    const tokenRow = Lexer.lexRow(input);
    // The Semantic types have more restrictions on where certain node types can appear.
    // We cast for now, but really we should have function that checks that the result of
    // editorParser.parse() follows those restrictions.
    const result = editorParser.parse(tokenRow.children) as Semantic.types.Node;

    return removeExcessParens(result);
};
