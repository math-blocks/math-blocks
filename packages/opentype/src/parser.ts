import { Blob } from 'buffer';

import { parseCFF } from './tables/cff';
import { parseCmap } from './tables/cmap';
import { parseMATH } from './tables/math';
import { parseHead } from './tables/head';

import { getGlyphMetrics } from './util';

import type { Font, TableRecord, TableDirectory } from './types';

const parseTag = (view: DataView, offset = 0): string => {
  return [
    view.getUint8(offset + 0),
    view.getUint8(offset + 1),
    view.getUint8(offset + 2),
    view.getUint8(offset + 3),
  ]
    .map((x) => String.fromCharCode(x))
    .join('');
};

const parseTableRecords = async (
  blob: Blob,
  offset: number,
  numTables: number,
): Promise<Record<string, TableRecord>> => {
  const RECORD_SIZE = 4 + 4 + 4 + 4;

  const buffer = await blob
    .slice(offset, offset + numTables * RECORD_SIZE)
    .arrayBuffer();
  const view = new DataView(buffer);

  const result: Record<string, TableRecord> = {};

  for (let i = 0; i < numTables; i++) {
    const offset = i * RECORD_SIZE;
    const tag = parseTag(view, offset);

    result[tag] = {
      tableTag: parseTag(view, offset),
      checksum: view.getUint32(offset + 4),
      offset: view.getUint32(offset + 8),
      length: view.getUint32(offset + 12),
    };
  }

  return result;
};

export const parseDirectory = async (blob: Blob): Promise<TableDirectory> => {
  const byteCount = 4 + 8;

  const buffer = await blob.slice(0, byteCount).arrayBuffer();
  const view = new DataView(buffer);

  const numTables = view.getUint16(4);

  const dir: TableDirectory = {
    sfntVersion: parseTag(view), // tag, 4 bytes
    numTables: numTables, // uint16
    searchRange: view.getUint16(6),
    entrySelector: view.getUint16(8),
    rangeShift: view.getUint16(10),
    tableRecords: await parseTableRecords(blob, 12, numTables),
  };

  return dir;
};

export const parse = async (blob: Blob): Promise<Font> => {
  const dir = await parseDirectory(blob);

  const headTableRecord = dir.tableRecords['head'];
  if (!headTableRecord) {
    throw new Error("No TableRecord for 'head' table");
  }
  const headBlob = blob.slice(
    headTableRecord.offset,
    headTableRecord.offset + headTableRecord.length,
  );
  const head = await parseHead(headBlob);

  // TODO: move the blob for each table record into the parseDirector result
  const cmapTableRecord = dir.tableRecords['cmap'];
  if (!cmapTableRecord) {
    throw new Error("No TableRecord for 'cmap' table");
  }
  const cmapBlob = blob.slice(
    cmapTableRecord.offset,
    cmapTableRecord.offset + cmapTableRecord.length,
  );
  const glyphIndexMap = await parseCmap(cmapBlob);

  const mathTableRecord = dir.tableRecords['MATH'];
  if (!mathTableRecord) {
    throw new Error("No TableRecord for 'MATH' table");
  }
  const mathBlob = blob.slice(
    mathTableRecord.offset,
    mathTableRecord.offset + mathTableRecord.length,
  );
  const math = await parseMATH(mathBlob, head.unitsPerEm);

  const cffTableRecord = dir.tableRecords['CFF '];
  if (!cffTableRecord) {
    throw new Error("No TableRecord for 'CFF ' table");
  }
  const cffBlob = blob.slice(
    cffTableRecord.offset,
    cffTableRecord.offset + cffTableRecord.length,
  );
  const cff = await parseCFF(cffBlob);

  const font: Font = {
    cff: {
      name: cff.name,
      topDict: cff.topDict,
    },
    head,
    math,
    getGlyph: cff.getGlyph,
    getGlyphID: (char: string): number => {
      const codePoint = char.codePointAt(0);
      if (!codePoint) {
        throw new Error(`Invalid char: ${char}`);
      }
      const glyphID = glyphIndexMap[codePoint];
      if (!glyphID) {
        throw new Error(`No glyphID in font for given codePoint: ${codePoint}`);
      }
      return glyphID;
    },
    getGlyphMetrics: (gid: number) => {
      const glyph = cff.getGlyph(gid);
      return getGlyphMetrics(glyph);
    },
  };

  return font;
};
