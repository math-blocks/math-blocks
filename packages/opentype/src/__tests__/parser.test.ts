import path from 'path';
import fs from 'fs';
import { Blob } from 'buffer';

import type { Font, TableDirectory } from '../types';

import { parse, parseDirectory } from '../parser';

describe('parseDirectory', () => {
  let dir: TableDirectory;

  beforeAll(async () => {
    const fontPath = path.join(__dirname, '../../../../assets/STIX2Math.otf');
    const buffer = fs.readFileSync(fontPath);
    const blob = new Blob([buffer]);

    dir = await parseDirectory(blob);
  });

  test('top-level properties', () => {
    expect(dir.sfntVersion).toEqual('OTTO');
    expect(dir.numTables).toEqual(13);
    expect(dir.searchRange).toEqual(128);
    expect(dir.entrySelector).toEqual(3);
    expect(dir.rangeShift).toEqual(80);
  });

  test('tableRecords', () => {
    expect(dir.tableRecords).toMatchSnapshot();
  });
});

describe('parse', () => {
  let font: Font;

  beforeAll(async () => {
    const fontPath = path.join(__dirname, '../../../../assets/STIX2Math.otf');
    const buffer = fs.readFileSync(fontPath);
    const blob = new Blob([buffer]);

    font = await parse(blob);
  });

  describe('CFF ', () => {
    test('name', () => {
      expect(font.cff.name).toEqual('STIXTwoMath');
    });

    test('topDict', () => {
      expect(font.cff.topDict).toMatchSnapshot();
    });
  });

  describe('math', () => {
    describe('MathContants', () => {
      test('constants should have expected values', () => {
        expect(font.math.constants).toMatchSnapshot();
      });
    });

    describe('MathVariants', () => {
      test('top-level variants properties', async () => {
        expect(font.math.variants).toMatchInlineSnapshot(`
                    Object {
                      "getHorizGlyphConstruction": [Function],
                      "getVertGlyphConstruction": [Function],
                      "horizGlyphCount": 47,
                      "horizGlyphCoverageOffset": 992,
                      "minConnectorOverlap": 100,
                      "vertGlyphCount": 118,
                      "vertGlyphCoverageOffset": 340,
                    }
                `);
      });

      test('vertical glyph construction for parenleft', () => {
        const gid = font.getGlyphID('(');

        const glyphConstruction =
          font.math.variants.getVertGlyphConstruction(gid);

        expect(glyphConstruction).toMatchSnapshot();
      });

      test('getVertGlyphConstruction is memoized', () => {
        const gid = font.getGlyphID('(');

        const construction1 = font.math.variants.getVertGlyphConstruction(gid);
        const construction2 = font.math.variants.getVertGlyphConstruction(gid);

        expect(construction1).toBe(construction2);
      });

      test('horizontal glyph construction for uni0303 (combining tilde)', () => {
        const gid = font.getGlyphID('\u0303');

        const glyphConstruction =
          font.math.variants.getHorizGlyphConstruction(gid);

        expect(glyphConstruction).toMatchSnapshot();
      });

      test('getHorizGlyphConstruction is memoized', () => {
        const gid = font.getGlyphID('\u0303');

        const construction1 = font.math.variants.getHorizGlyphConstruction(gid);
        const construction2 = font.math.variants.getHorizGlyphConstruction(gid);

        expect(construction1).toBe(construction2);
      });
    });

    describe('MathGlyphInfo', () => {
      test("')' should be an extended glyph", () => {
        const gid = font.getGlyphID(')');

        expect(font.math.glyphInfo.isExtendedShape(gid)).toBe(true);
      });

      test("'+' should not be an extended glyph", () => {
        const gid = font.getGlyphID('+');

        expect(font.math.glyphInfo.isExtendedShape(gid)).toBe(false);
      });
    });
  });

  describe('glyphs', () => {
    test('name', () => {
      const gid = font.getGlyphID('(');
      const glyph = font.getGlyph(gid);

      expect(glyph.name).toEqual('parenleft');
    });

    test('getGlyph is memoized', () => {
      const gid = font.getGlyphID('(');
      const glyph1 = font.getGlyph(gid);
      const glyph2 = font.getGlyph(gid);

      expect(glyph1).toBe(glyph2);
    });

    test('metrics', () => {
      const gid = font.getGlyphID('(');
      const glyph = font.getGlyph(gid);

      expect(glyph.metrics).toEqual({
        advance: 357,
      });
    });

    test('getGlyphMetrics', () => {
      const gid = font.getGlyphID('(');
      const glyphMetrics = font.getGlyphMetrics(gid);

      expect(glyphMetrics).toEqual({
        advance: 357,
        bearingX: 45,
        bearingY: 736,
        height: 932,
        width: 282,
      });
    });

    test("metrics and glyphMetrics 'advance' match", () => {
      const gid = font.getGlyphID('(');
      const glyphMetrics = font.getGlyphMetrics(gid);
      const glyph = font.getGlyph(gid);

      expect(glyphMetrics.advance).toEqual(glyph.metrics.advance);
    });

    test('path', () => {
      const gid = font.getGlyphID('(');
      const glyph = font.getGlyph(gid);

      expect(glyph.path).toEqual([
        { type: 'M', x: 45, y: 270 },
        { type: 'C', x: 311, x1: 45, x2: 150, y: -196, y1: 47, y2: -112 },
        { type: 'L', x: 327, y: -170 },
        { type: 'C', x: 127, x1: 181, x2: 127, y: 270, y1: -71, y2: 66 },
        { type: 'C', x: 327, x1: 127, x2: 181, y: 710, y1: 474, y2: 611 },
        { type: 'L', x: 311, y: 736 },
        { type: 'C', x: 45, x1: 150, x2: 45, y: 270, y1: 652, y2: 493 },
        { type: 'Z' },
      ]);
    });
  });
});
