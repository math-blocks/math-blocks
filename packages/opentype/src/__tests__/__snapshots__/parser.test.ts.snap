// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse CFF  topDict 1`] = `
Object {
  "BlueFuzz": 1,
  "BlueScale": 0.039625,
  "BlueShift": 6,
  "BlueValues": Array [
    -15,
    15,
    469,
    16,
    25,
    20,
    73,
    13,
    14,
    17,
    8,
    13,
    36,
    15,
  ],
  "CharStrings": 56085,
  "CharstringType": 2,
  "Copyright": "Arie de Ruiter, who in 1995 was Head of Information Technology Development at Elsevier Science, made a proposal to the STI Pub group, an informal group of publishers consisting of representatives from the American Chemical Society \\\\(ACS\\\\), American Institute of Physics \\\\(AIP\\\\), American Mathematical Society \\\\(AMS\\\\), American Physical Society \\\\(APS\\\\), Elsevier, and Institute of Electrical and Electronics Engineers \\\\(IEEE\\\\). De Ruiter encouraged the members to consider development of a series of Web fonts, which he proposed should be called the Scientific and Technical Information eXchange, or STIX, Fonts. All STI Pub member organizations enthusiastically endorsed this proposal, and the STI Pub group agreed to embark on what has become a twelve-year project. The goal of the project was to identify all alphabetic, symbolic, and other special characters used in any facet of scientific publishing and to create a set of Unicode-based fonts that would be distributed free to every scientist, student, and other interested party worldwide. The fonts would be consistent with the emerging Unicode standard, and would permit universal representation of every character. With the release of the STIX fonts, de Ruiter's vision has been realized.",
  "Encoding": 0,
  "ExpansionFactor": 0.06,
  "FamilyName": "STIX Two Math",
  "FontBBox": Array [
    -978,
    -1641,
    3072,
    2627,
  ],
  "ForceBold": false,
  "FullName": "STIX Two Math",
  "ItalicAngle": 0,
  "LanguageGroup": 0,
  "Notice": "Copyright \\\\(c\\\\) 2001-2016 by the STI Pub Companies, consisting of the American Chemical Society, the American Institute of Physics, the American Mathematical Society, the American Physical Society, Elsevier, Inc., and The Institute of Electrical and Electronic Engineers, Inc.  Portions copyright \\\\(c\\\\) 1998-2003 by MicroPress, Inc.  Portions copyright \\\\(c\\\\) 1990 by Elsevier, Inc.  All rights reserved.  Portions copyright \\\\(c\\\\) Copyright 2010, 2012, 2014 Adobe Systems Incorporated.",
  "OtherBlues": Array [
    -235,
    15,
  ],
  "PaintType": 0,
  "Private": Array [
    691667,
    41,
  ],
  "StdHW": 29,
  "StdVW": 83,
  "StemSnapH": Array [
    29,
    12,
  ],
  "StemSnapV": Array [
    83,
    9,
  ],
  "StrokeWidth": 0,
  "UnderlinePosition": -103,
  "UnderlineThickness": 51,
  "Weight": "Normal",
  "charset": 55096,
  "defaultWidthX": 701,
  "fontMatrix": Array [
    0.001,
    0,
    0,
    0.001,
    0,
    0,
  ],
  "initialRandomSee": 0,
  "isFixedPitch": false,
  "nominalWidthX": 571,
  "version": "001.001",
}
`;

exports[`parse math MathContants constants should have expected values 1`] = `
Object {
  "accentBaseHeight": 0.48,
  "axisHeight": 0.258,
  "delimitedSubFormulaMinHeight": 1.325,
  "displayOperatorMinHeight": 1.8,
  "flattenedAccentBaseHeight": 0.656,
  "fractionDenomDisplayStyleGapMin": 0.15,
  "fractionDenominatorDisplayStyleShiftDown": 0.64,
  "fractionDenominatorGapMin": 0.068,
  "fractionDenominatorShiftDown": 0.585,
  "fractionNumDisplayStyleGapMin": 0.15,
  "fractionNumeratorDisplayStyleShiftUp": 0.64,
  "fractionNumeratorGapMin": 0.068,
  "fractionNumeratorShiftUp": 0.585,
  "fractionRuleThickness": 0.068,
  "lowerLimitBaselineDropMin": 0.67,
  "lowerLimitGapMin": 0.135,
  "mathLeading": 0.15,
  "overbarExtraAscender": 0.068,
  "overbarRuleThickness": 0.068,
  "overbarVerticalGap": 0.175,
  "radicalDegreeBottomRaisePercent": 55,
  "radicalDisplayStyleVerticalGap": 0.17,
  "radicalExtraAscender": 0.078,
  "radicalKernAfterDegree": -0.335,
  "radicalKernBeforeDegree": 0.065,
  "radicalRuleThickness": 0.068,
  "radicalVerticalGap": 0.085,
  "scriptPercentScaleDown": 70,
  "scriptScriptPercentScaleDown": 55,
  "skewedFractionHorizontalGap": 0.35,
  "skewedFractionVerticalGap": 0.068,
  "spaceAfterScript": 0.04,
  "stackBottomDisplayStyleShiftDown": 0.69,
  "stackBottomShiftDown": 0.385,
  "stackDisplayStyleGapMin": 0.3,
  "stackGapMin": 0.15,
  "stackTopDisplayStyleShiftUp": 0.78,
  "stackTopShiftUp": 0.47,
  "stretchStackBottomShiftDown": 0.59,
  "stretchStackGapAboveMin": 0.068,
  "stretchStackGapBelowMin": 0.068,
  "stretchStackTopShiftUp": 0.8,
  "subSuperscriptGapMin": 0.15,
  "subscriptBaselineDropMin": 0.16,
  "subscriptShiftDown": 0.21,
  "subscriptTopMax": 0.368,
  "superscriptBaselineDropMax": 0.23,
  "superscriptBottomMaxWithSubscript": 0.38,
  "superscriptBottomMin": 0.12,
  "superscriptShiftUp": 0.36,
  "superscriptShiftUpCramped": 0.252,
  "underbarExtraDescender": 0.068,
  "underbarRuleThickness": 0.068,
  "underbarVerticalGap": 0.175,
  "upperLimitBaselineRiseMin": 0.3,
  "upperLimitGapMin": 0.135,
}
`;

exports[`parse math MathVariants horizontal glyph construction for uni0303 (combining tilde) 1`] = `
Object {
  "glyphAssembly": null,
  "mathGlyphVariantRecords": Array [
    Object {
      "advanceMeasurement": 322,
      "variantGlyph": 742,
    },
    Object {
      "advanceMeasurement": 574,
      "variantGlyph": 1405,
    },
    Object {
      "advanceMeasurement": 1003,
      "variantGlyph": 1406,
    },
    Object {
      "advanceMeasurement": 1496,
      "variantGlyph": 1407,
    },
    Object {
      "advanceMeasurement": 1932,
      "variantGlyph": 1408,
    },
    Object {
      "advanceMeasurement": 2385,
      "variantGlyph": 1409,
    },
  ],
}
`;

exports[`parse math MathVariants vertical glyph construction for parenleft 1`] = `
Object {
  "glyphAssembly": Object {
    "italicsCorrection": Object {
      "deviceOffset": 0,
      "value": 0,
    },
    "partRecords": Array [
      Object {
        "endConnectorLength": 250,
        "fullAdvance": 1273,
        "glyphID": 4862,
        "partsFlags": 0,
        "startConnectorLength": 0,
      },
      Object {
        "endConnectorLength": 1000,
        "fullAdvance": 1252,
        "glyphID": 4861,
        "partsFlags": 1,
        "startConnectorLength": 1000,
      },
      Object {
        "endConnectorLength": 0,
        "fullAdvance": 1273,
        "glyphID": 4860,
        "partsFlags": 0,
        "startConnectorLength": 250,
      },
    ],
  },
  "mathGlyphVariantRecords": Array [
    Object {
      "advanceMeasurement": 933,
      "variantGlyph": 1064,
    },
    Object {
      "advanceMeasurement": 1187,
      "variantGlyph": 1301,
    },
    Object {
      "advanceMeasurement": 1427,
      "variantGlyph": 1302,
    },
    Object {
      "advanceMeasurement": 1667,
      "variantGlyph": 1303,
    },
    Object {
      "advanceMeasurement": 1907,
      "variantGlyph": 1304,
    },
    Object {
      "advanceMeasurement": 2145,
      "variantGlyph": 1305,
    },
    Object {
      "advanceMeasurement": 2385,
      "variantGlyph": 1306,
    },
    Object {
      "advanceMeasurement": 2625,
      "variantGlyph": 1307,
    },
    Object {
      "advanceMeasurement": 2865,
      "variantGlyph": 1308,
    },
    Object {
      "advanceMeasurement": 3101,
      "variantGlyph": 1309,
    },
    Object {
      "advanceMeasurement": 3341,
      "variantGlyph": 1310,
    },
    Object {
      "advanceMeasurement": 3581,
      "variantGlyph": 1311,
    },
    Object {
      "advanceMeasurement": 3821,
      "variantGlyph": 1312,
    },
  ],
}
`;

exports[`parseDirectory tableRecords 1`] = `
Object {
  "CFF ": Object {
    "checksum": 2978540771,
    "length": 691708,
    "offset": 220,
    "tableTag": "CFF ",
  },
  "GDEF": Object {
    "checksum": 356002236,
    "length": 66,
    "offset": 691928,
    "tableTag": "GDEF",
  },
  "GPOS": Object {
    "checksum": 2118371111,
    "length": 23252,
    "offset": 691996,
    "tableTag": "GPOS",
  },
  "GSUB": Object {
    "checksum": 2636999998,
    "length": 7360,
    "offset": 715248,
    "tableTag": "GSUB",
  },
  "MATH": Object {
    "checksum": 4056928493,
    "length": 23292,
    "offset": 722608,
    "tableTag": "MATH",
  },
  "OS/2": Object {
    "checksum": 1836143976,
    "length": 96,
    "offset": 745900,
    "tableTag": "OS/2",
  },
  "cmap": Object {
    "checksum": 3520504363,
    "length": 29314,
    "offset": 745996,
    "tableTag": "cmap",
  },
  "head": Object {
    "checksum": 348148424,
    "length": 54,
    "offset": 775312,
    "tableTag": "head",
  },
  "hhea": Object {
    "checksum": 203759218,
    "length": 36,
    "offset": 775368,
    "tableTag": "hhea",
  },
  "hmtx": Object {
    "checksum": 2942074728,
    "length": 22166,
    "offset": 775404,
    "tableTag": "hmtx",
  },
  "maxp": Object {
    "checksum": 363286528,
    "length": 6,
    "offset": 797572,
    "tableTag": "maxp",
  },
  "name": Object {
    "checksum": 1587078982,
    "length": 10608,
    "offset": 797580,
    "tableTag": "name",
  },
  "post": Object {
    "checksum": 4290052147,
    "length": 32,
    "offset": 808188,
    "tableTag": "post",
  },
}
`;
